BM/tools/pcie/pcie_check.c

```diff
@@ -0,0 +1,1225 @@
+		printf("open /dev/mem failed!\n");
+		return -1;
+	}
+
+	if (BASE_ADDR == 0)
+		find_bar();
+
+	addr = BASE_ADDR | (bus << 20) | (dev << 15) | (fun << 12);
+	ptrdata = mmap(NULL, LEN_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, addr);
+
+	printf("Offset addr:%lx, *ptrdata:%x, LEN_SIZE:%lx\n", addr, *ptrdata, LEN_SIZE);
+	if ((*ptrdata != ptr_content) && (*ptrdata != 0)) {
+		printf("%02x:%02x.%01x:", bus, dev, fun);
+
+		if (((check_list >> 1) & 0x1) == 1) {
+			for (offset = 0; offset < 64; offset++) {
+				if (offset % 4 == 0)
+					printf("\n%02x: ", offset * 4);
+				printf("%02x ", (u8)(*(ptrdata + offset) >> 0));
+				printf("%02x ", (u8)(*(ptrdata + offset) >> 8));
+				printf("%02x ", (u8)(*(ptrdata + offset) >> 16));
+				printf("%02x ", (u8)(*(ptrdata + offset) >> 24));
+			}
+			if (is_pcie == 1) {
+				for (offset = 64; offset < 1024; offset++) {
+					if (offset % 4 == 0)
+						printf("\n%02x: ", offset * 4);
+					printf("%02x ", (u8)(*(ptrdata + offset) >> 0));
+					printf("%02x ", (u8)(*(ptrdata + offset) >> 8));
+					printf("%02x ", (u8)(*(ptrdata + offset) >> 16));
+					printf("%02x ", (u8)(*(ptrdata + offset) >> 24));
+				}
+			}
+			printf("\n");
+		}
+		if (is_pcie == 1)
+			check_pcie(ptrdata);
+		else
+			check_pci(ptrdata);
+	} else {
+		printf("*ptrdata:%x, which is 0 or %x, ptrdata:%p, return\n",
+		       *ptrdata, ptr_content, ptrdata);
+	}
+	munmap(ptrdata, LEN_SIZE);
+	close(fd);
+	return 0;
+}
+
+int recognize_pcie(u32 *ptrdata)
+{
+	int loop_num = 0;
+	u8 nextpoint;
+	u32 *ptrsearch;
+
+	is_pcie = 0;
+	/* 0x34/4 is capability pointer in PCI */
+	nextpoint = (u8)(*(ptrdata + PCI_CAP_START / 4));
+
+	if (nextpoint == 0)
+		return 0;
+
+	ptrsearch = ptrdata + nextpoint / 4;
+	while (1) {
+		/* 0x10 means PCIE capability */
+		if ((u8)(*ptrsearch) == 0x10) {
+			is_pcie = 1;
+			break;
+		}
+		if ((u8)(*ptrsearch) == 0xff) {
+			printf("*ptrsearch:%x offset is 0xff, ptrsearch:%p, ptrdata:%p\n",
+			       *ptrsearch, ptrsearch, ptrdata);
+			return 2;
+		}
+
+		/* no PCIE find */
+		if ((u8)((*ptrsearch) >> 8) == 0x00)
+			break;
+		if (loop_num >= 16)
+			break;
+		/* next capability */
+		ptrsearch = ptrdata + ((u8)(((*ptrsearch) >> 8)	& 0x00ff)) / 4;
+		loop_num++;
+	}
+	return 0;
+}
+
+int scan_pci(void)
+{
+	// Must be 64bit address for 64bit OS!
+	u64 addr = 0;
+	u32 bus, dev, fun;
+	// Must 32bit for data check!
+	u32 *ptrdata = malloc(sizeof(unsigned long) * 4096), *ps;
+	u8 nextpoint;
+
+	int fd;
+
+	fd = open("/dev/mem", O_RDWR);
+
+	if (fd < 0) {
+		printf("open /dev/mem failed!\n");
+		return -1;
+	}
+	printf("fd=%d open /dev/mem successfully.\n", fd);
+
+	ptrdata = &ptr_content;
+	for (bus = 0; bus < MAX_BUS; ++bus) {
+		for (dev = 0; dev < MAX_DEV; ++dev) {
+			for (fun = 0; fun < MAX_FUN; ++fun) {
+				addr = BASE_ADDR | (bus << 20) | (dev << 15) | (fun << 12);
+
+				ptrdata = mmap(NULL, LEN_SIZE, PROT_READ | PROT_WRITE,
+					       MAP_SHARED, fd, addr);
+
+				if (ptrdata == (void *)-1) {
```

Comment:
It can be more readable if using MAP_FAILED instead of -1 to check the ptrdata.

Reply:
Yes, will update it.

Thanks!